Create REST Controllers for Users-Roles CRUD operations + Implement add service logic for base entities (User)

iryna-pantsialeyeva opened the pull request:
Please review my PR

iryna-pantsialeyeva committed:
add basic CRUD (without U) for User (controller + service)
update entities' @OneToMany relationships
add custom methods (findByName) in repositories
add custom exceptions + @ExceptionHandler methods
update schema.sql in accordance with updated entities
update data.sql

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/entity/Game.java:
```diff
@@ -34,13 +36,19 @@ public Game(String gameName, String description, String picPath) {
-    @JoinColumn(name = "producers_id", referencedColumnName = "id")
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "producers_id", nullable = true)
+    @OnDelete(action = OnDeleteAction.NO_ACTION)
```
не обязательно вообще (я б убрал)

iryna-pantsialeyeva replied:
Сделано

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/entity/Game.java:
```diff
@@ -34,13 +36,19 @@ public Game(String gameName, String description, String picPath) {
+    @OnDelete(action = OnDeleteAction.NO_ACTION)
     private Producer producer;
+
+    public void addSale(Sale sale) {
+        sales.add(sale);
+        sale.setGame(this);
+    }
```
не надо так делать. Убери это пожалуйста

iryna-pantsialeyeva replied:
Согласна, что можно без addSale(). Достаточно геттера для получения получения хэшсета и вызова на них метода add().

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/entity/User.java:
```diff
@@ -37,4 +37,12 @@ public User(String userName, String password, String email) {
     @OneToMany(mappedBy = "user")
     private Set<Sale> sales = new HashSet<>();
 
+    public void addRole(Role role) {
+        roles.add(role);
+    }
+
+    public void addSale(Sale sale) {
+        sales.add(sale);
+        sale.setUser(this);
+    }
```
вот это - плохо так делать, если уж хибернейт используем

iryna-pantsialeyeva replied:
Согласна, что можно без них. Достаточно геттера для получения получения хэшсетов и вызова на них метода add().

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/entity/Sale.java:
```diff
@@ -19,12 +19,13 @@ public class Sale {
-    @ManyToOne(cascade = CascadeType.ALL)
+    @ManyToOne(fetch = FetchType.LAZY)
     @JoinColumn(name = "games_id")
+    @OnDelete(action = OnDeleteAction.CASCADE)
```
лишнее

iryna-pantsialeyeva replied:
При удалении игры мы хотим, чтобы удалялась sale, без аннотацию падает с ошибкой, т.к. есть связанная запись в сейле.

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/exception/ObjectAlreadyExistException.java:
```diff
@@ -0,0 +1,7 @@
+package info.freeit.boarderpicker.exception;
+
+public class ObjectAlreadyExistException extends Exception {
```
я бы убрал и юзал стандартный IllegalArgumentException

iryna-pantsialeyeva replied:
Сделано.

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/exception/ObjectNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package info.freeit.boarderpicker.exception;
+
+public class ObjectNotFoundException extends Exception {
```
замени это все на какой-то более general эксепшн. типа GamesPersistanceException или чет такое

iryna-pantsialeyeva replied:
ObjectNotFoundException заменено на ObjectPersistanceException

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/service/Impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Override
+    public List<UserDTO> getAllUsers() {
+        List<User> users = Streamable.of(userRepository.findAll()).toList();
```
а тебе точно нужна поточная обработка? Давай просто сразу список из репозитория достанем, а?

iryna-pantsialeyeva replied:
Точно нужна. Я хочу вернуть List, а на коллекции типа Iterable нельзя вызвать stream чтобы преобразовать в List объектов dto.

RomanKashenok replied:
тогда давай так:
return Streamable.of(userRepository.findAll()).stream().map(user -> UserDTO.fromUser(user)).toList();

iryna-pantsialeyeva replied:
Сделано

iryna-pantsialeyeva replied:
сделано

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/service/Impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+    @Override
+    public void deleteUserByID(int id) throws ObjectNotFoundException {
+        if (userRepository.existsById(id)) {
+            userRepository.deleteById(id);
```
ты можешь сразу удалять по айди не делая два запроса в базу

iryna-pantsialeyeva replied:
Можно сразу удалять.  Но из описания метода deleteById() : Если энтити с таким id не будет найден, это будет проигнорировано "silently". Я хотела, чтобы клиент видел, что он ввел несуществующий id (реализовано с помощью выброса кастомного эксепшена).

iryna-pantsialeyeva committed:
add U in basic CRUD for User (controller + service)
update in accordance with Roman's notes

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/exception/ObjectPersistenceException.java:
```diff
@@ -0,0 +1,7 @@
+package info.freeit.boarderpicker.exception;
+
+public class ObjectPersistenceException extends Exception {
```
Давай в следующих коммитах сделаем его рантайм и уберем его пробрасывания в коде

iryna-pantsialeyeva replied:
Убрала вместе с папкой exception

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/controller/UserController.java:
```diff
@@ -29,17 +29,22 @@ public List<UserDTO> getAllUsers() {
 
     @DeleteMapping("/delete")
-    public void deleteUser(@RequestParam int id) throws ObjectNotFoundException {
+    public void deleteUser(@RequestParam int id) throws ObjectPersistenceException {
```
Убери это как сделаешь исключение рантайм

iryna-pantsialeyeva replied:
Сделано

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/service/Impl/UserServiceImpl.java:
```diff
@@ -28,28 +27,38 @@ public List<UserDTO> getAllUsers() {
 
     @Override
-    public UserDTO getUserByID(int userID) throws ObjectNotFoundException {
+    public UserDTO getUserByID(int userID) throws ObjectPersistenceException {
```
убери ObjectPersistenceException как сделаешь рантайм

iryna-pantsialeyeva replied:
сделано

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/service/UserService.java:
```diff
@@ -2,17 +2,18 @@
     List<UserDTO> getAllUsers();
 
-    UserDTO getUserByID(int userID) throws ObjectNotFoundException;
+    UserDTO getUserByID(int userID) throws ObjectPersistenceException;
```
убери как сделаешь рантайм

iryna-pantsialeyeva replied:
сделано

RomanKashenok commented on src/main/java/info/freeit/boarderpicker/controller/UserController.java:
```diff
@@ -0,0 +1,50 @@
+        return userService.getUserByID(userID);
+    }
+
+    @PostMapping("/save")
```
убери слова эти /save, /delete, /update - это и так из методов понятно

iryna-pantsialeyeva replied:
сделано

iryna-pantsialeyeva committed:
replace ObjectPersistenceException with RuntimeException (+delete package "exception")
delete "throws ObjectPersistenceException" from method signature in controller and service layer
delete "save/update/delete" from mapping
shorten implementation of getAllUsers() method
